### DISCUSSION TITLE
Next.js hot reload not working for Socket.IO?

### DISCUSSION THREAD

**Discussion by SimHoZebs** [vercel/next.js#42657] [Help]
I have a Next.js project that has the simplest Socket.IO implementation set up. Below is the code.

```js
//  pages/index.tsx

let socket: Socket;
const Home: NextPage = () => {
  useEffect(() => {
    async function socketInit() {
      //start server
      await fetch("http://localhost:3000/api/socket");

      // connects to the socket
      socket = io();
      socket.on("connect", () => {
        console.log("hello");
      });
    }

    socketInit();
  }, []);

  return (
    <button
      onClick={() => {
        socket.emit("test");
      }}
    >
      hello
    </button>
  );
};
```

```js
//  pages/api/socket.ts

export default function handler(
  req: NextApiRequest,
  res: Res
) {
  if (res.socket.server.io) { res.end(); return; }

  const io = new IOServer(res.socket.server);
  res.socket.server.io = io;
  io.on('connection', socket => {
    socket.on('test', () => {
      console.log("1");  //Changing it to "2" doesn't do anything until dev is restarted.
    });
  });

  res.end();
}
```

For some reason, the `listener` in the server would not update from hot reload. Restarting the dev is the only way. Is there something I'm missing?

**Comment by icyJoseph** [comment]
I think the issue is that even though the hot reloader picks up on the change, there's a process started by the IOServer which just keeps on going.

I tweaked it like this:

```ts
let io: IOServer;

export default function handler(req: NextApiRequest, res: Res) {
  if (res.socket.server.io) {
    if (io === res.socket.server.io) {
      console.log("Socket is already running");
      return res.end();
    }
  }
  console.log("Socket is initializing");
  io = new IOServer(res.socket.server);

  res.socket.server.io = io;

  io.on("connection", (socket) => {
    socket.on("input-change", (msg) => {
      console.log("hola");

      socket.broadcast.emit("update-input", msg);
    });
  });

  return res.end();
}
```

And also made some modifications to the client side code:

```tsx
import { SocketAddress } from "net";
import React, { useEffect, useRef, useState } from "react";
import { io, Socket } from "socket.io-client";

const ServerTest = () => {
  const socketRef = useRef<Socket>();

  useEffect(() => {
    async function socketInit(signal: AbortSignal) {
      // signals the server to create a socket
      await fetch("http://localhost:3000/api/socket", { signal });

      // connects to the socket
      socketRef.current = io("http://localhost:3000");

      const socket = socketRef.current;

      socket.on("connect", () => {
        console.log("Connected, connection id is ", socket.id);
      });

      socket.on("update-input", (msg) => {
        console.log("server just said holy");
      });
    }

    const controller = new AbortController();

    socketInit(controller.signal).catch((e) => {
      if (controller.signal.aborted) {
        console.debug("User aborted the request");
      }
    });

    return () => {
      controller.abort();
      if (socketRef.current) {
        socketRef.current.close();
      }
    };
  }, []);

  return (
    <button
      onClick={() => {
        socketRef.current?.emit("input-change", "hi");
      }}
    >
      ServerTest
    </button>
  );
};

export default ServerTest;
```

And now it kinda works... you still need the client to "disconnect" from the socket, and next time it reconnects it'll pick up on the new instance of the handler.

## Caveat

Web-sockets and Next.js are sort of at-odds, because for example when deploying to Vercel, API handlers last maximum 10 seconds on free-tier because these are made to be stateless serverless functions. So a Socket process running in one of them will be terminated by the runtime. 

For toy projects it might be fine to use the Socket within an API route, but production, as soon as the runtime optimises API handlers as serverless functions, all bets are off.

The Next.js examples used to have this one: https://github.com/vercel/next.js/tree/442fbfaa4d4e8205cf10c7e5d27b8f51b7a4bdc6/examples/with-socket.io, and you can see it uses a custom server to run the socket. Look at this answer https://github.com/vercel/next.js/discussions/14950#discussioncomment-983725.

**Reply by SimHoZebs** [reply]
Thank you for the response! As for the serverless function, I was thinking of deploying it somewhere other than Vercel, within the options they recommended: https://vercel.com/guides/publish-and-subscribe-to-realtime-data-on-vercel But thinking about it again, it seems that I'm using serverless functions for the wrong goal. I might make a custom server.

I just have one question about your implementation:

> you still need the client to "disconnect" from the socket, and next time it reconnects it'll pick up on the new instance of the handler.

My initial implementation returned different socket ID whenever I refreshed the browser. Hence, I assumed that I didn't need to explicitly disconnect the client from it like you've done. Could you explain what the difference is?

Thanks again!

**Reply by icyJoseph** [reply]
I reckon I left a bunch of code I used to debug this situation. The window unload stuff is not really needed :) I was after trying to kill the running process in some fashion. So you can ignore that! I'll remove it from the code.

What I meant by `you still need the client to "disconnect" from the socket` is that you just need to somehow connect again, for example by refreshing the browser, or by navigating to another route and then using the clean up function.
